streams/shared/src/main/scala/zio/stream/StreamPure.scala

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stream
+
+import zio._
+
+private[stream] trait StreamPure[+A] extends ZStream[Any, Nothing, A] { self =>
```

Comment:
I would not call this `StreamPure`, but maybe, `StreamEffect`.

Also, I'd say the following:

 - We should permit `() => A` to throw `EndOfStreamException`, defined package-private so only we can use it
 - We should permit `() => A` to throw StreamFailedException(e: E)`, defined package-private so only we can use it
 - We should use the above to implement end of stream & failure, for both "pure streams" and synchronous effectful streams

Done properly, this will permit streams as fast as Scala collections, but lazy and potentially infinite; albeit only for blocking / synchronous, and pure streams.

Async streams are always gonna require chunking for performance.